### PR TITLE
Add convenience methods for use in Ethereum libraries

### PR DESCRIPTION
**Pull request by JamesSmartCell**
- Add constructor to initialise a uint256 from a hex string.
- Add export_bits to uint256 to export fixed length 32 byte vector.
- Add export_bits_truncate to export fitted byte vector.

### CODE REVIEW

File: uint128_t.cpp [hunk 1]
```diff
@@ -20,6 +20,46 @@ uint128_t::uint128_t(uint128_t && rhs)
     }
 }
 
+uint128_t::uint128_t(const char *s)
```
Comment (calccrypto) [inline uint128_t.cpp]: Why a `char *`? Why not `std::string`?

File: uint128_t.cpp [hunk 2]
```diff
@@ -20,6 +20,46 @@ uint128_t::uint128_t(uint128_t && rhs)
...
+	while (hv != 0xFF && count < 16)
+	{
+		val = (val << 4) | hv;
+		hv = HexToInt(&s[count]);
+		count++;
+	}
+	return val;
+}
+
+uint8_t uint128_t::HexToInt(const char *s) {
```
Comment (calccrypto) [inline uint128_t.cpp]: Is a pointer necessary here?

File: uint128_t.cpp [hunk 3]
```diff
@@ -256,6 +296,24 @@ uint128_t & uint128_t::operator*=(const uint128_t & rhs){
...
+	ret->push_back(static_cast<uint8_t>(*val >> 48));
+	ret->push_back(static_cast<uint8_t>(*val >> 40));
+	ret->push_back(static_cast<uint8_t>(*val >> 32));
+	ret->push_back(static_cast<uint8_t>(*val >> 24));
+	ret->push_back(static_cast<uint8_t>(*val >> 16));
+	ret->push_back(static_cast<uint8_t>(*val >> 8));
+	ret->push_back(static_cast<uint8_t>(*val));
+}
+
+void uint128_t::export_bits(std::vector<uint8_t> *ret) const
```
Comment (calccrypto) [inline uint128_t.cpp]: Pass by reference

File: uint128_t.cpp [hunk 4]
```diff
@@ -20,6 +20,46 @@ uint128_t::uint128_t(uint128_t && rhs)
     }
 }
 
+uint128_t::uint128_t(const char *s)
+{
```
Comment (calccrypto) [inline uint128_t.cpp]: Also can do conversion from any base rather than just base16.

File: uint128_t.include [hunk 5]
```diff
@@ -59,6 +60,7 @@ class uint128_t{
         uint128_t();
         uint128_t(const uint128_t & rhs);
         uint128_t(uint128_t && rhs);
+		uint128_t(const char *val);
```
Comment (calccrypto) [inline uint128_t.include]: Please uses 4 spaces instead of tabs.

File: uint256_t.cpp [hunk 6]
```diff
@@ -1,4 +1,6 @@
 #include "uint256_t.build"
+#include <vector>
+#include <string.h>
```
Comment (calccrypto) [inline uint256_t.cpp]: `#include <cstring>` for C++ headers.

File: uint128_t.cpp (hunk 1 above)
Comment (calccrypto) [reply uint128_t.cpp]: If `char *` is preferable to `std::string`, maybe have the `std::string` function call the `char *` version?

File: uint128_t.cpp [hunk 7]
```diff
@@ -20,6 +20,46 @@ uint128_t::uint128_t(uint128_t && rhs)
...
+	if (s[1] == 'x')
+		s += 2;
+	else if (*s == 'x')
+		s++;
+
+	UPPER = ConvertToUint64(s);
+	LOWER = ConvertToUint64(s+16);
+}
+
+uint64_t uint128_t::ConvertToUint64(const char *s)
```
Comment (calccrypto) [inline uint128_t.cpp]: This function can be `const`.

File: uint128_t.cpp (hunk 2 above)
Comment (calccrypto) [inline uint128_t.cpp]: This function can be const.

**Comment by JamesSmartCell** [comment]
Thankyou for your review! My C++ is very rusty (circa 2004/5) as you can see. I've made the changes you requested. Mostly, the unusual choices to do things that way were dictated by trying to use as little memory as possible, but in hindsight the compiler will be producing the same code and that style is better.

I did some other tidying up so it looks the same as your code formatting style. Hopefully it looks a bit better now. Please let me know if there's any other changes I can do and thanks again for your review. This library has been a life saver, as Ethereum uses uint256 for a lot of values and has some very large numbers for handling the small 'Wei' unit.

File: uint128_t.cpp [hunk 8]
```diff
@@ -20,6 +20,47 @@ uint128_t::uint128_t(uint128_t && rhs)
     }
 }
 
+uint128_t::uint128_t(std::string s) {
```
Comment (calccrypto) [inline uint128_t.cpp]: This can be passed by const reference.

File: uint128_t.include [hunk 9]
```diff
@@ -59,6 +60,8 @@ class uint128_t{
         uint128_t();
         uint128_t(const uint128_t & rhs);
         uint128_t(uint128_t && rhs);
+        uint128_t(std::string s);
```
Comment (calccrypto) [inline uint128_t.include]: This can be passed by const reference.

File: uint256_t.cpp [hunk 10]
```diff
@@ -24,6 +26,33 @@ uint256_t::uint256_t(uint256_t && rhs)
     }
 }
 
+uint256_t::uint256_t(const std::string s){
```
Comment (calccrypto) [inline uint256_t.cpp]: This can be passed by const reference.

File: uint256_t.include [hunk 11]
```diff
@@ -55,6 +55,8 @@ class uint256_t{
         uint256_t();
         uint256_t(const uint256_t & rhs);
         uint256_t(uint256_t && rhs);
+        uint256_t(const std::string s);
```
Comment (calccrypto) [inline uint256_t.include]: This can be passed by const reference.

**Comment by calccrypto** [comment]
@JamesSmartCell Can I also bother you to make a pull request for the `uint128_t` changes in the [`uint128_t` repo](https://github.com/calccrypto/uint128_t)?

**Comment by JamesSmartCell** [comment]
Updated to use reference and deduplicate the two constructor paths.

**Pull request merged**